package/MDAnalysis/analysis/rms.py

```diff
@@ -398,125 +416,100 @@ def __init__(self, traj, reference=None, select='all',
 
-    def run(self, start=None, stop=None, step=None,
-            mass_weighted=None, ref_frame=None):
-        """Perform RMSD analysis on the trajectory.
 
-        A number of parameters can be changed from the defaults. The
-        result is stored as the array :attr:`RMSD.rmsd`.
+    def _prepare(self):
+        self._n_atoms = self.mobile_atoms.n_atoms
 
-        Parameters
-        ----------
-        start, stop, step : int (optional)
-            start and stop frame index with step size
-        mass_weighted : bool (optional)
-            overwrite object default to do a mass-weighted RMSD fit
-        ref_frame : int
-             frame index to select frame from *reference*
-        """
-        if mass_weighted is None:
-            mass_weighted = self.mass_weighted
-        if ref_frame is None:
-            ref_frame = self.ref_frame
-
-        natoms = self.traj_atoms.n_atoms
-        trajectory = self.universe.trajectory
-        traj_atoms = self.traj_atoms
-
-        if mass_weighted:
-            # if performing a mass-weighted alignment/rmsd calculation
-            weight = self.ref_atoms.masses / self.ref_atoms.masses.mean()
-            weight = weight.astype(np.float64)
-        else:
-            weight = None
+        self._weights = ((self.ref_atoms.masses / self.ref_atoms.masses.mean())
+                        if self.mass_weighted else None)
+
+        # reference centre of mass system, no longer 1 based indices
+        current_frame = self.reference.trajectory.ts.frame
 
-        # reference centre of mass system
-        current_frame = self.reference.trajectory.ts.frame - 1
         try:
             # Move to the ref_frame
             # (coordinates MUST be stored in case the ref traj is advanced
             # elsewhere or if ref == mobile universe)
-            self.reference.trajectory[ref_frame]
-            ref_com = self.ref_atoms.center_of_mass()
+            self.reference.trajectory[self.ref_frame]
+            self._ref_com = self.ref_atoms.center_of_mass()
             # makes a copy
-            ref_coordinates = self.ref_atoms.positions - ref_com
-            if self.groupselections_atoms:
-                groupselections_ref_coords_64 = [
-                    self.reference.select_atoms(*s['reference']).positions.astype(np.float64) for s in
+            self._ref_coordinates = self.ref_atoms.positions - self._ref_com
+            if self._groupselections_atoms:
+                self._groupselections_ref_coords_64 = [(self.reference.
+                    select_atoms(*s['reference']).
+                    positions.astype(np.float64)) for s in
                     self.groupselections]
         finally:
             # Move back to the original frame
             self.reference.trajectory[current_frame]
-        ref_coordinates_64 = ref_coordinates.astype(np.float64)
 
-        # allocate the array for selection atom coords
-        traj_coordinates = traj_atoms.positions.copy()
```

Comment:
If you allocate an array for mobile coordinates you could avoid recreating it in `_single_step()`.
```python
self._mobile_coordinates64 = self.mobile_atoms.positions.copy().astype(np.float64)
```

and then (numpy will cast to float64 because the type of `self._mobile_coordinates` is fixed) do inplace ops:
```python
self._mobile_coordinates64[:] = self.mobile_atoms.positions
self._mobile_coordinates64 -= mobile_com
```